### ISSUE TITLE
Before submitting a PR: XUnit test runner that can be used directly

### ISSUE THREAD

**Issue by breki** [fscheck/FsCheck#509]
Hi,

As [discussed on Twitter](https://twitter.com/breki74/status/1203632900069044224), I needed an XUnit runner that actually writes all of its output using XUnit's `ITestOutputHelper` and that can be used directly, not just through `Property` attributes, so I implemented one by (mostly) cannibalizing existing code from FsCheck: https://github.com/breki/demeton/blob/master/Demeton.Tests/BetterFsCheckRunner.fs

Here's a sample on how it is used: https://github.com/breki/demeton/blob/master/Demeton.Tests/PropertiesHelp.fs#L46

While I could add the whole code as a PR, I think it would be a better idea to discuss how best to integrate it - do I just merge the code with the existing XUnit runner or implement it separately?

**Comment by kurtschelfthout** [comment]
Thanks for considering contributing to FsCheck!

I think this should go in the FsCheck.Xunit project/namespace. There's already an `XunitRunner` public type [there](https://github.com/fscheck/FsCheck/blob/master/src/FsCheck.Xunit/PropertyAttribute.fs#L19), so unfortunately we can't change that into your implementation, so it's going to have to be a new type. If possible it may be worth deprecating `XunitRunner` and using your runner implementation in `PropertyAttribute` also.

Does that make sense? Feel free to ping me if you get stuck or want me to take a more detailed look.

**Comment by breki** [comment]
Yes, it makes sense :-). I'll try to find a way to prevent any breaking changes and code duplication.
 
One thing I do remember when I was working on this back in December is that I had to copy & paste some functions to be used in the new runner because the original ones were private and thus inaccessible to the outside world.

In any case, I will give you write access to my fork so you can change things you don't like before the PR is created.

**Comment by kurtschelfthout** [comment]
Sounds good!

**Comment by breki** [comment]
Are there any docs describing the prerequisites needed to run the build?

I'm getting this warning:
```
warning NU1701: Package 'FSharp.Core 4.0.0.1' was restored using '.NETPortable,Version=v0.0,Profile=Profile259' 
instead of the project target framework '.NETStandard,Version=v1.0'. 
This package may not be fully compatible with your project.
```

Here's the result of `dotnet --info` on my machine: https://gist.github.com/breki/ee4e5d851995a0feb1c74fd33b6d46a3

**Comment by kurtschelfthout** [comment]
That warning is expected you can ignore it.

**Comment by breki** [comment]
Should I ignore these warnings as welI?
```
D:\src\breki-FsCheck\examples\FsCheck.Examples\FsCheck.Examples.fsproj.metaproj 
: warning MSB4078: The project file "examples\FsCheck.Examples\FsCheck.Examples.fsproj" is not supported by MSBuild and cannot be built.
```

Also, when I compile the code in VS (both in Debug and Release configurations), it fails with these kinds of errors:
```
D:\src\breki-FsCheck\src\FsCheck\Arbitrary.fs(510,40): error FS0001: 
The type 'Numerics.BigInteger' does not support a conversion to the type 'decimal'
1>D:\src\breki-FsCheck\src\FsCheck\Runner.fs(257,17): error FS0039: 
The value or constructor 'printf' is not defined. Maybe you want one of the following:   Printf   fprintf   sprintf   fprintfn   PrintfFormat
```

**Comment by kurtschelfthout** [comment]
> Should I ignore these warnings as welI?

No, that's unexpected.

> Also, when I compile the code in VS (both in Debug and Release configurations), it fails with these kinds of errors:

Did you try running `build.cmd` in the root of the repo first? Do these errors also appear there or is it only in VS?

**Comment by breki** [comment]
> Did you try running `build.cmd` in the root of the repo first? Do these errors also appear there or is it only in VS?

Yes, I ran `build.cmd` first. And these errors appear in VS only.

**Comment by breki** [comment]
I tried building in Rider, maybe this error from the build log can shed more light:

```
"D:\src\breki-FsCheck\src\FsCheck\FsCheck.fsproj" (build target) (1) ->
(ResolvePackageAssets target) -> 
 D:\src\breki-FsCheck\src\FsCheck\FsCheck.fsproj : error NU1202: Package FSharp.Core 4.2.3 is not compatible with netstandard1.0 (.NETStandard,Version=v1.0). Package FSharp.Core 4.2.3 supports:
D:\src\breki-FsCheck\src\FsCheck\FsCheck.fsproj : error NU1202:   - net45 (.NETFramework,Version=v4.5)
D:\src\breki-FsCheck\src\FsCheck\FsCheck.fsproj : error NU1202:   - netstandard1.6 (.NETStandard,Version=v1.6)
```

**Comment by kurtschelfthout** [comment]
Strange that it works on command line but not in any IDE. I’ll try in a fresh clone tonight I guess. Perhaps there’s a step missing in build.cmd or something like that.

**Comment by breki** [comment]
> Strange that it works on command line but not in any IDE. I’ll try in a fresh clone tonight I guess. Perhaps there’s a step missing in build.cmd or something like that.

Or I am missing something installed on my machine (an older SDK or something)? In any case, let me know when you get the chance.

**Comment by kurtschelfthout** [comment]
I updated VS and rebuild, and that works for me, see output below.

I've also added the output of `dotnet --info`. Do you have .NET Framework installed? It still targets net452 and .NET Portable, besides netstandard. Sorry these dependencies have accrued over time and I guess I've always kept my machines up to date and working. Certainly intend to document this when we get to the bottom of it.

Can you paste the output of build.cmd somewhere? Perhaps that gives me a clue.

```
1>------ Rebuild All started: Project: FsCheck, Configuration: Debug Any CPU ------
1>FsCheck -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck\bin\Debug\netstandard1.0\FsCheck.dll
1>FsCheck -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck\bin\Debug\net452\FsCheck.dll
1>FsCheck -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck\bin\Debug\netstandard1.6\FsCheck.dll
1>FsCheck -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck\bin\Debug\netstandard2.0\FsCheck.dll
2>------ Rebuild All started: Project: FsCheck.Xunit, Configuration: Debug Any CPU ------
3>------ Rebuild All started: Project: FsCheck.NUnit, Configuration: Debug Any CPU ------
4>------ Rebuild All started: Project: FsCheck.Examples, Configuration: Debug Any CPU ------
5>------ Rebuild All started: Project: FsCheck.CSharpExamples, Configuration: Debug Any CPU ------
5>FsCheck.CSharpExamples -> C:\Users\Kurt\Projects\FsCheck\FsCheck\examples\FsCheck.CSharpExamples\bin\Debug\net452\FsCheck.CSharpExamples.exe
6>------ Rebuild All started: Project: FsCheck.MsTest.Examples, Configuration: Debug Any CPU ------
3>FsCheck.NUnit -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck.NUnit\bin\Debug\net452\FsCheck.NUnit.dll
2>FsCheck.Xunit -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck.Xunit\bin\Debug\net452\FsCheck.Xunit.dll
4>FsCheck.Examples -> C:\Users\Kurt\Projects\FsCheck\FsCheck\examples\FsCheck.Examples\bin\Debug\net452\FsCheck.Examples.exe
3>FsCheck.NUnit -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck.NUnit\bin\Debug\netstandard1.6\FsCheck.NUnit.dll
6>FsCheck.MsTest.Examples -> C:\Users\Kurt\Projects\FsCheck\FsCheck\examples\FsCheck.MsTest.Examples\bin\Debug\net452\FsCheck.MsTest.Examples.dll
3>FsCheck.NUnit -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck.NUnit\bin\Debug\netstandard2.0\FsCheck.NUnit.dll
7>------ Rebuild All started: Project: FsCheck.NUnit.Examples, Configuration: Debug Any CPU ------
8>------ Rebuild All started: Project: FsCheck.NUnit.CSharpExamples, Configuration: Debug Any CPU ------
2>FsCheck.Xunit -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck.Xunit\bin\Debug\netstandard1.6\FsCheck.Xunit.dll
8>FsCheck.NUnit.CSharpExamples -> C:\Users\Kurt\Projects\FsCheck\FsCheck\examples\FsCheck.NUnit.CSharpExamples\bin\Debug\net452\FsCheck.NUnit.CSharpExamples.dll
2>FsCheck.Xunit -> C:\Users\Kurt\Projects\FsCheck\FsCheck\src\FsCheck.Xunit\bin\Debug\netstandard2.0\FsCheck.Xunit.dll
9>------ Rebuild All started: Project: FsCheck.Test, Configuration: Debug Any CPU ------
10>------ Rebuild All started: Project: FsCheck.XUnit.CSharpExamples, Configuration: Debug Any CPU ------
11>------ Rebuild All started: Project: CSharp.DocSnippets, Configuration: Debug Any CPU ------
7>FsCheck.NUnit.Examples -> C:\Users\Kurt\Projects\FsCheck\FsCheck\examples\FsCheck.NUnit.Examples\bin\Debug\net452\FsCheck.NUnit.Examples.dll
11>CSharp.DocSnippets -> C:\Users\Kurt\Projects\FsCheck\FsCheck\docs\csharp\bin\Debug\net452\CSharp.DocSnippets.exe
10>FsCheck.XUnit.CSharpExamples -> C:\Users\Kurt\Projects\FsCheck\FsCheck\examples\FsCheck.XUnit.CSharpExamples\bin\Debug\net452\FsCheck.Xunit.CSharpExamples.dll
9>FsCheck.Test -> C:\Users\Kurt\Projects\FsCheck\FsCheck\tests\FsCheck.Test\bin\Debug\net452\FsCheck.Test.dll
========== Rebuild All: 11 succeeded, 0 failed, 0 skipped ==========
```

```
>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.101
 Commit:    b377529961

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.101\

Host (useful for support):
  Version: 3.1.1
  Commit:  a1388f194c

.NET Core SDKs installed:
  2.2.401 [C:\Program Files\dotnet\sdk]
  3.0.100 [C:\Program Files\dotnet\sdk]
  3.1.101 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by breki** [comment]
Comparing your `dotnet --info` with mine, I don't see any major differences that would cause problems. 

I have these .NET Framework versions installed (Chocolatey packages):
```
netfx-4.5.1-devpack 4.5.50932
netfx-4.5.2-devpack 4.5.5165101.20180721
netfx-4.6.1-devpack 4.6.01055.00
netfx-4.7.2-devpack 4.7.2.20190225
netfx-4.8 4.8.0.20190930
netfx-4.8-devpack 4.8.0.20190930
```

How do you build releases? Directly on your machine or do you perhaps have a Docker image or a GitHub action? That would help me determine what I'm missing on my system.

The log of `build.cmd` is here: https://gist.github.com/breki/21df920d03c94d7cc03097569b8ef928

**Comment by kurtschelfthout** [comment]
I think the problem is it is using an old version of MSBuild.
```
c:\Windows\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe  "D:\src\breki-FsCheck\FsCheck.sln" "/t:Rebuild" "/m:8" "/nodeReuse:False" "/p:RestorePackages=False" "/p:Configuration=Release"
c:\Windows\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe  "D:\src\breki-FsCheck\FsCheck.sln" "/t:Rebuild" "/m:8" "/nodeReuse:False" "/p:RestorePackages=False" "/p:Configuration=Release"
Microsoft (R) Build Engine version 4.8.3752.0
```

I'm looking at this - likely a problem in the build script or Fake or some such. Not clear what changed though...this definitely used to pick MSBuild 15 or higher.

> How do you build releases? Directly on your machine or do you perhaps have a Docker image or a GitHub action? That would help me determine what I'm missing on my system.

Actual releases I build on my machine, for convenience since I have the NuGet key. That said artifacts are built on AppVeyor: https://ci.appveyor.com/project/kurtschelfthout/fscheck

**Comment by kurtschelfthout** [comment]
So I'm still not quite sure why this is happening, but it's indeed due to using an old version of MSBuild.

I've changed line 117 of `build.fsx` to:

```
    |> MSBuild.runRelease (fun par -> { par with MaxCpuCount = Some (Some Environment.ProcessorCount)
                                                 ToolPath = @"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe" }) 
                            "" "Rebuild"
```

And this appears to work around the problem, although it still gives me a bogus (I think!) "Target Build failed" at the end of the build. 

It does appear that for me in Visual Studio at least it uses the correct MSBuild so I'm not getting any errors there. For me the errors only appear when using `build.cmd`.

Could you try this workaround and see if it helps?

I'm going to investigate this further - perhaps I finally need to bite the bullet and update to the new fake runner which also runs on dotnet core.

**Comment by breki** [comment]
Changing the `build.fsx` as you suggested did work in FAKE, but it does not help in Visual Studio.

However, I did some investigating, and found a solution, but I'm not sure whether it is acceptable for FsCheck.

I found that all of the solution projects  (except one) used `FSharp.Core` v4.2.3, while `FsCheck` project used v4.0.0.1. So I tried to update it to 4.2.3, but it complained that .NET Standard 1.0 is not supported by 4.2.3.

So the next step I did was to remove .NET Standard 1.0 support from `FsCheck.fsproj`, so instead of:
```xml
<TargetFrameworks>net452;netstandard2.0;netstandard1.6;netstandard1.0</TargetFrameworks>
```
I replaced i with
```xml
<TargetFrameworks>net452;netstandard2.0;netstandard1.6</TargetFrameworks>
```
 
Then I updated `FSharp.Core` to 4.2.3 and now the solution builds both in Visual Studio and in Fake (even with the original `build.fsx` script), without any warnings or errors.

Is ending support for .NET Standard 1.0 acceptable? 

BTW, I did some reading about [.NET Standard versions](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) and it says .NET Standard 1.6 (the minimum that would now be supported in FsCheck) targets .NET Framework 4.6.1.

**Comment by kurtschelfthout** [comment]
Ending support for 1.0 is something I’ve been putting off as long as possible but I do want to get there. Same with older framework versions. It causes a bunch of build related complexities and slowdown - just standard 1.6 and 2.0 would be great. 

However it would have to go in a 3.0 release bc it’s not backwards compatible, and I’d like to do that together with signing the assemblies and removing some deprecated API. I don’t want to block you on that. 

What I suggest you do is you prepare your PR in the way you just described, and then I’ll do some work on it afterwards to either put it on the 3.0 roadmap or make it work with 2.x (the latter being my preferred outcome).

**Comment by breki** [comment]
> What I suggest you do is you prepare your PR in the way you just described, and then I’ll do some work on it afterwards to either put it on the 3.0 roadmap or make it work with 2.x (the latter being my preferred outcome).

That's fine. I'll be on travelling in the next few days, but hopefully I'll have something to show by the end of the next week.

**Comment by kurtschelfthout** [comment]
You may want to retry after pulling latest from master. I just did a pretty comprehensive build tooling overhaul. https://github.com/fscheck/FsCheck/pull/504

**Comment by breki** [comment]
> You may want to retry after pulling latest from master. I just did a pretty comprehensive build tooling overhaul. #504

I can't seem to update FSharp.Core to v4.2.3 now using NuGet now, has the way the 3rd party libraries are referenced been changed?

**Comment by kurtschelfthout** [comment]
Hm, no, still uses paket as before. Dependencies have not changed either.

**Comment by breki** [comment]
Hmm I'm confused now, because this definitively worked before :-). I haven't used paket before, so I probably don't know what I'm doing. 

Anyway, I wanted to get rid of 4.0.0.1 reference, so I removed `portable` paket group, but now I have a different problem: a lot of tests are failing with this kind of error:

```
System.IO.FileLoadException
Could not load file or assembly 'FSharp.Core, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
   at FsCheck.Experimental.Test.StateMachine.check generated commands()
```

To be honest, I'm starting to think the best idea would be for me to wait for your build system upgrade before continuing to work on this PR, because I don't have enough knowledge about all the required configurations FsCheck has to support, otherwise I'll massacre the paket dependencies.

Or you could simply use the my runner file and integrate it to FsCheck the way you think it's best: https://github.com/breki/FsCheck/blob/improved-xunit-runner/src/FsCheck.Xunit/BetterFsCheckRunner.fs

**Comment by kurtschelfthout** [comment]
Fair enough, this kind of stuff can be somewhat exasperating. 

Hard to debug from here as well, works on AppVeyor and Travis CI so unclear what’s missing. To build you should only need .NET Core 3.1.101 now; I didn’t get the test runs working on dotnet test yet, so those still use the netfx/mono based runner.

**Comment by breki** [comment]
Maybe the problem is not what's missing, but what else is there (I have other .NET Core versions installed together with 3.1.101, and also different .NET Frameworks). I also suspect there is more than one issue here and that's why it's hard to debug. Certainly one issue is with FAKE using the wrong MSBuild version (at least on two machines I tested this with).

**Comment by kurtschelfthout** [comment]
Released in 2.16.3.
